### PR TITLE
Minimal proposal spec

### LINKED ISSUE #8: A more minimal version
TLDR:

We don't add signals into the language.

Details:

Add an internal slot:

```js
[[RevocationSignal]]: Record { revoked: boolean, callbacks: list of Abstract Closure }
```
These `callbacks` should be synchronously invoked.

1. Host object `AbortSignal` is having this slot.
2. new Proxy:
    1. Let _signal_ be options?.signal
    2. If _signal_ does not have a `[[RevocationSignal]]` internal slot, throw a TypeError.
    3. Let _revocable_ be %Proxy.revocable%(_target_, _handler_)
    4. Let _revokeFunction_ be _revocable_.revoke
    5. Let _closure_ be a new Abstract Closure that captures _revokeFunction_ with the following steps:
        1. Call _revokeFunction_()
    6. Push _closure_ to _signal_.[[RevocationSignal]].callbacks
    8. Return _revocable_.proxy

### PR DESCRIPTION
**Pull request by Jack-Works**
close #8 

image

image

image

image

### CODE REVIEW

**Comment by acutmore** [comment]
Only having the `isRevoked` getter and not a way to _listen_ for the revoke call seems like it would add a delay to when the `target` and `handler` references are dropped allowing them to be collected?

**Comment by ajvincent** [comment]
I think having `Proxy.revocable()` **not** always return `{proxy, revoke}` is something we can not ask for.  Though it may seem weird, I can foresee someone out there wanting to still revoke an individual proxy, while letting hundreds of others live.  Plus, it's a change in what the function actually returns from a generic dictionary to a Proxy.

So the algorithm at 1.3 step 3b has to go, and the "Else" branch at 1.4 has to be unconditionally run, I think.

**Review by ajvincent — CHANGES_REQUESTED** [review]
I don't think we can change the return signature of Proxy.revocable.

**Comment by Jack-Works** [comment]
The current semantics allows the engine to GC all targets & handlers of the proxy.

> I think having Proxy.revocable() not always return {proxy, revoke} is something we can not ask for.

What about changing `Proxy.revocable` to return `{ proxy: T, revoke?: Function }`? I prefer adding this functionality to the `Proxy.revocable` instead of `new Proxy` because it has better semantics.

**Comment by Jack-Works** [comment]
> Though it may seem weird, I can foresee someone out there wanting to still revoke an individual proxy, while letting hundreds of others live.

Oh... This is definitely reasonable. So the correct choice is: accept it on both `Proxy.revocable` and `new Proxy` (as 3rd argument)

**Comment by ajvincent** [comment]
> The current semantics allows the engine to GC all targets & handlers of the proxy.
> 
> > I think having Proxy.revocable() not always return {proxy, revoke} is something we can not ask for.
> 
> What about changing `Proxy.revocable` to return `{ proxy: T, revoke?: Function }`? I prefer adding this functionality to the `Proxy.revocable` instead of `new Proxy` because it has better semantics.

This proposal is definitely aiming at `Proxy.revocable`, and considering `new Proxy` in #6.

**Comment by Jack-Works** [comment]
spec updated.

**Review by ajvincent — APPROVED** [review]
You've answered my objection.  I'll wait for the others to review before merging, for now.

**Comment by ajvincent** [comment]
I still have a concern about directly referencing AbortController as the mass-revocation signal basis.

**Comment by Jack-Works** [comment]
Maybe we can follow https://github.com/tc39/proposal-cancellation/issues/31 as our foundation.

**Comment by ajvincent** [comment]
There's a conflict.  Can you please update this PR to account for that?

**Comment by Jack-Works** [comment]
@ajvincent done

File: index.html
Comment (ajvincent) [inline index.html:2552]: I'm not sure if you updated the spec text so step 4h, which sets the `[[RevocationSignal]]` to null despite step 3a.  

I'll bet it's standard for the spec editing code to put all of this on one line, but it's really hard on the eyes to have to scroll so far in to fix a bug on one line.

I will merge with this bug in, and I ask you to write a follow-up PR to fix the bug.

Per #6, I'd also like a follow-up to modify `new Proxy()` as well.

Comment (ajvincent) [reply index.html:2552]: Ah, I see it's generated from spec.emu below.  I didn't know that.

**Review by ajvincent — APPROVED** [review]

**Pull request merged**
